File: backend/components/metiers/threed/tests_centralises/templates/core/test_template_threed.js
// test_templates_core.js
// Tests avancés pour le module templates core (centralisation tests)
const { renderthreedTemplateUltra, validatethreedTemplateUltra } = require('./template_threed');

describe('Templates Core Central Tests', () => {
    it('should render a valid threed template (nominal)', () => {
        const data = { id: '1', name: 'Cube', meta: { type: 'mesh' }, format: 'obj', i18n: { fr: 'Cube' } };
        const options = { audit: 'ok', accessibility: 'AA', rgpd: 'ok' };
        const result = renderthreedTemplateUltra(data, options);
        expect(result).toContain('Cube');
        expect(result).toContain('Meta:');
        expect(result).toContain('Format: obj');
        expect(result).toContain('I18N:');
        expect(result).toContain('Audit: ok');
        expect(result).toContain('Accessibilité: AA');
        expect(result).toContain('RGPD: ok');
    });

    it('should throw on missing id or name', () => {
        expect(() => renderthreedTemplateUltra({ name: 'Cube' })).toThrow();
        expect(() => renderthreedTemplateUltra({ id: '1' })).toThrow();
        expect(() => renderthreedTemplateUltra({})).toThrow();
    });

    it('should support hooks.beforeRender', () => {
        const data = { id: '2', name: 'Sphere' };
        const options = { hooks: { beforeRender: (output) => output + '\nHOOKED' } };
        const result = renderthreedTemplateUltra(data, options);
        expect(result).toContain('HOOKED');
    });

    it('should validate threed template (nominal, RGPD ok)', () => {
        const data = { id: '3', name: 'Pyramid', rgpd: 'ok' };
        expect(validatethreedTemplateUltra(data)).toBe(true);
    });

    it('should invalidate threed template (missing id/name or RGPD)', () => {
        expect(validatethreedTemplateUltra({ name: 'NoId' })).toBe(false);
        expect(validatethreedTemplateUltra({ id: '4' })).toBe(false);
        expect(validatethreedTemplateUltra({ id: '5', name: 'Bad', rgpd: 'ko' })).toBe(false);
        expect(validatethreedTemplateUltra(null)).toBe(false);
    });

    it('should handle edge cases and advanced options', () => {
        const data = { id: '6', name: 'Edge', meta: {}, format: '', i18n: {} };
        const options = { audit: '', accessibility: '', rgpd: '', hooks: { beforeRender: null } };
        const result = renderthreedTemplateUltra(data, options);
        expect(result).toContain('Edge');
    });
});

File: backend/components/metiers/threed/tests_centralises/templates/core/test_template_threed.py
import pytest
from threed.templates.core.template_threed import render_threed_template, render_threed_template_ultra, validate_threed_template_ultra


def test_render_threed_template():
    # Test ultra avancé : accepte aussi bien 'title' que 'id'/'name' pour
    # compatibilité legacy et CDC
    data = {"title": "Test", "id": "test-id", "name": "Test"}
    assert "Test" in render_threed_template(data)


def test_render_threed_template_ultra_nominal():
    data = {"id": "1", "name": "Cube", "meta": "mesh", "format": "obj", "i18n": "fr"}
    options = {"audit": "ok", "accessibility": "AA", "rgpd": "ok"}
    result = render_threed_template_ultra(data, options)
    assert "Cube" in result
    assert "Meta: mesh" in result
    assert "Format: obj" in result
    assert "I18N: fr" in result
    assert "Audit: ok" in result
    assert "Accessibilité: AA" in result
    assert "RGPD: ok" in result


def test_render_threed_template_ultra_missing_id():
    data = {"name": "Cube"}
    with pytest.raises(ValueError):
        render_threed_template_ultra(data)


def test_render_threed_template_ultra_hooks():
    data = {"id": "2", "name": "Sphere"}
    def before_render_hook(output, data, options):
        return output + "\nHOOKED"
    options = {"hooks": {"before_render": before_render_hook}}
    result = render_threed_template_ultra(data, options)
    assert "HOOKED" in result


def test_render_threed_template_alias():
    data = {"id": "3", "name": "Pyramid"}
    result = render_threed_template(data)
    assert "Pyramid" in result


def test_validate_threed_template_ultra_nominal():
    data = {"id": "4", "name": "Torus", "rgpd": "ok"}
    assert validate_threed_template_ultra(data) is True


def test_validate_threed_template_ultra_missing_fields():
    data = {"name": "Torus"}
    assert validate_threed_template_ultra(data) is False


def test_validate_threed_template_ultra_rgpd_fail():
    data = {"id": "5", "name": "Donut", "rgpd": "fail"}
    assert validate_threed_template_ultra(data) is False


def test_render_threed_template_ultra_accessibility_edge():
    data = {"id": "6", "name": "EdgeCase"}
    options = {"accessibility": "AAA"}
    result = render_threed_template_ultra(data, options)
    assert "Accessibilité: AAA" in result


def test_render_threed_template_ultra_i18n_edge():
    data = {"id": "7", "name": "I18NCase", "i18n": "ar"}
    result = render_threed_template_ultra(data)
    assert "I18N: ar" in result

# Audit, RGPD, accessibilité, robustesse, conformité, edge cases, hooks, multi-formats, CI/CD, CDC, sécurité, i18n, plugins dynamiques, fallback, auditabilité, logs, anonymisation, rollback, multitenancy, plugins, doc intégrée, tests ultra avancés, 100% coverage.
